### PR TITLE
Allow any quantity if variant is set to on demand

### LINKED ISSUE #7297: "On demand" items cannot have their quantities changed in backoffice orders
## Description

If a variant is set to "on demand", changing the quantity in the admin edit order screen is not allowed.

## Expected Behavior

Changing the quantity should be allowed

## Actual Behaviour

Clicking the arrows has no effect

## Steps to Reproduce

1. Create an order with a variant set to "on demand"
2. Visit admin edit order page
3. Click the pencil to change the quantity of the line item for that variant
4. Click the arrows to change the quantity
5. Nothing happens

## Animated Gif/Screenshot

LUagSwfaDE

## Workaround

None? 
## Severity
Could probably be an S2

## Your Environment

* Version used:
* Browser name and version:
* Operating System and version (desktop or mobile):

### PR DESCRIPTION
**Pull request by andrewpbrett**
#### What? Why?

Closes #7297 

The admin order form had some logic to prevent more quantity than was available from being ordered. This checks if the variant is set to be on demand and allows any quantity to be ordered if so. 

#### What should we test?

See issue gif: Changing the quantity of a line item with an on demand variant. 

#### Release notes

Fixed a bug in the admin edit order screen where on-demand variants could not have their quantity increased. 

Changelog Category: User facing changes

### CODE REVIEW

File: app/views/spree/admin/orders/_shipment_manifest.html.haml [hunk 1]
```diff
@@ -14,7 +14,9 @@
           = "#{count} x #{t(state.humanize.downcase, scope: [:spree, :shipment_states], default: [:missing, "none"])}"
       - unless shipment.shipped?
         %td.item-qty-edit.hidden
-          = number_field_tag :quantity, item.quantity, :min => 0, :class => "line_item_quantity", :size => 5, :max => item.variant.on_hand + item.quantity
+          - html_options = { :min => 0, :class => "line_item_quantity", :size => 5 }
+          - html_options.merge!({ :max => item.variant.on_hand + item.quantity }) unless item.variant.on_demand
+          = number_field_tag :quantity, item.quantity, html_options
```
Comment (sauloperez) [inline app/views/spree/admin/orders/_shipment_manifest.html.haml:19]: we're stuffing quite some Ruby logic into the view already. I think this better fit into a helper method like `quantity_field_tag` :ok_hand:

File: spec/features/admin/order_spec.rb [hunk 2]
```diff
@@ -146,6 +146,27 @@ def new_order_with_distribution(distributor, order_cycle)
...
+      expect(page).to have_input(:quantity)
+      fill_in(:quantity, with: 1000)
+      find("a.save-item").click
+    end
+
+    within("tr.stock-item", text: order.products.first.name) do
+      expect(page).to have_text("1000 x")
+    end
+    expect(order.reload.line_items.first.quantity).to eq(1000)
+  end
```
Comment (sauloperez) [inline spec/features/admin/order_spec.rb:168]: sweeet :clap:

File: app/views/spree/admin/orders/_shipment_manifest.html.haml (hunk 1 above)
Comment (andrewpbrett) [reply app/views/spree/admin/orders/_shipment_manifest.html.haml]: Good idea! f32dc8c50d2c4b810bb880b3850ebec68bd346cf

**Review by Matt-Yorkley — APPROVED** [review]

File: app/helpers/spree/admin/orders_helper.rb [hunk 3]
```diff
@@ -120,6 +120,14 @@ def event_link(event)
...
                        data: { confirm: confirm_message })
       end
+
+      def quantity_field_tag(manifest_item)
+        html_options = { min: 0, class: "line_item_quantity", size: 5 }
+        unless manifest_item.variant.on_demand
+          html_options.merge!(max: manifest_item.variant.on_hand + manifest_item.quantity)
+        end
+        number_field_tag :quantity, manifest_item.quantity, html_options
+      end
```
Comment (sauloperez) [inline app/helpers/spree/admin/orders_helper.rb:130]: :clap: :clap:

**Review by sauloperez — APPROVED** [review]

**Comment by filipefurtad0** [comment]
Hey @andrewpbrett ,
nice catch!

This seems not to happen with all the variants. I manage to reproduce it by creating an `on demand` weight product and variant, from scratch, and adding it to the order.

After this PR, it was possible to change amount from this line item again:

Peek 2021-04-08 11-07

Good to go :+1:

**Pull request merged**